pkg/apis/experimental/validation/validation.go

```diff
@@ -333,3 +345,130 @@ func ValidateJobSpecUpdate(oldSpec, spec experimental.JobSpec) errs.ValidationEr
+				allErrs = append(allErrs, errs.NewFieldInvalid("host", host[0], errMsg))
+			}
+			if len(host) == 2 {
+				hostPort, err := strconv.Atoi(host[1])
+				if err != nil {
+					allErrs = append(allErrs, errs.NewFieldInvalid("host", ih.Host, "string following `:` delimiter must represent a port number."))
+				}
+				if !utilvalidation.IsValidPortNum(hostPort) {
+					allErrs = append(allErrs, errs.NewFieldInvalid("host", ih.Host, portRangeErrorMsg))
+				}
+			}
+			if len(host) > 2 {
+				allErrs = append(allErrs, errs.NewFieldInvalid("host", ih.Host, "Only a single `:` is allowed in the host, and this must delimit a single port"))
+			}
+		}
+		allErrs = append(allErrs, validateIngressRule(&ih.IngressRule).Prefix("ingressRule")...)
+	}
+	return allErrs
+}
+
+func validateIngressRule(ingressRule *experimental.IngressRule) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+	if ingressRule.HTTP != nil {
+		allErrs = append(allErrs, validateHTTPIngressRule(ingressRule.HTTP).Prefix("http")...)
+	}
+	return allErrs
+}
+
+func validateHTTPIngressRule(httpIngressRule *experimental.HTTPIngressRule) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+	if len(httpIngressRule.Paths) == 0 {
+		allErrs = append(allErrs, errs.NewFieldRequired("paths"))
+	}
+	for _, rule := range httpIngressRule.Paths {
+		if len(rule.Path) > 0 {
+			if !strings.HasPrefix(rule.Path, "/") {
+				allErrs = append(allErrs, errs.NewFieldInvalid("path", rule.Path, "path must begin with /"))
+			}
+			// TODO: More draconian path regex validation.
+			// Path needs to be a valid regex. This is the basic requirement.
+			// In addition to this any characters not allowed in a path per
+			// RFC 3986 section-3.3 cannot appear as a literal in the regex.
+			// Consider the example: http://host/valid?#bar, this is a valid
+			// regex that matches valid#bar, which isn't a valid path, because
+			// the path terminates at the first ? or #. A more sophisticated
+			// form of validation would detect that the user is confusing url
+			// regexes with path regexes.
+			_, err := regexp.CompilePOSIX(rule.Path)
+			if err != nil {
+				allErrs = append(allErrs, errs.NewFieldInvalid("path", rule.Path, "httpIngressRule.path needs to be a valid regex."))
+			}
+		}
+		allErrs = append(allErrs, validateIngressBackend(&rule.Backend).Prefix("backend")...)
+	}
+	return allErrs
+}
+
+// validateIngressBackend tests if a given backend is valid.
+func validateIngressBackend(backend *experimental.IngressBackend) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+
+	// All backends must reference a single local service by name, and a single service port by name or number.
+	if len(backend.ServiceRef.Name) == 0 {
+		return append(allErrs, errs.NewFieldRequired("serviceRef.name"))
+	} else if ok, errMsg := apivalidation.ValidateServiceName(backend.ServiceRef.Name, false); !ok {
```

Comment:
Something that we don't do, but probably should, that walks the line between validation and REST: cross-object link verification and refcounting.

@lavalamp for ideas, or just to put it on the radar.  We're doing more and more cross-object stuff.  Should we validate that the target of the link exists?  Should we prevent deletion of the linkee as long as the refcount is non-zero?  We'll have to keep back-links.  Complex, but the whole API is getting complex.